### ISSUE TITLE
isDirty property not changed when using Controller and shouldUnregister: false

### ISSUE THREAD

**Issue by Mhal007** [react-hook-form/react-hook-form#3459]
**Describe the bug**
When the `Controller` is used for a field that is an array (of numbers, strings, does not matter) and the `shouldUnregister` is `false`, then for some cases the form state is not refreshed after changes.

**To Reproduce**

1. Go to https://codesandbox.io/s/react-hook-form-controller-template-forked-qdf1e?file=/src/index.js
2. Open up the console, it's correctly stating that a form is not dirty
2. Click on 'Remove 2'
3. Check the console, nothing happened, still not dirty

**Codesandbox link (Required)**
https://codesandbox.io/s/react-hook-form-controller-template-forked-qdf1e?file=/src/index.js

**Expected behavior**
Form should be dirty if any of the values are changed

**Desktop:**
- OS: Manjaro (Linux Arch)
- Browser: Chrome
- Version: 86.0.4240.183 (Official Build) (64-bit)

**Additional context**
When you remove '1' it works fine, but removing '2' does not cause the from state to be changed.
Btw, I'm using `shouldUnregister: false` as I need it for my 'real' form that spans across multiple tabs, this is just a simplified example.

P.S. May not be needed, but here's a bit more complicated example, showing more specifically what I'm trying to do:
https://codesandbox.io/s/focused-architecture-up76i?file=/src/App.js

I have a list of string elements and some buttons for adding/removing them, as above. Removing 'test string' in this case does not cause the state to be changed.

**Comment by bluebill1049** [comment]
`shouldUnregister:false` doesn't update `isDirty` when the component is unmounted, because the input state remains the same only apply when the input value is changed.

**Comment by Mhal007** [comment]
@bluebill1049 Thanks for your response. I don't quite understand what you mean though. The only thing being unmounted here is the `div` containing the `button`. Even weirder is that pressing on 'Remove 1' works fine, but 'Remove 2' is not. It's not a case where an input in unmounted, I basically only have some buttons and texts. I was assuming that calling `onChange` is changing the state of the form (and again, it is, in case of the first button, but not the second one).

Let me know what I'm missing.

**Comment by bluebill1049** [comment]
I see, with button 2 after you change the value, the value is the same as the default value 

```
const defaultListValue = [1, 2];
[1, 2]
```

**Comment by Mhal007** [comment]
@bluebill1049 More specifically, button 2 just does not do anything. If you click on button 1 first, your state is [2], if you then press button 2, nothing changes, the state is still [2] (at least that's what the console says).

So do you think that it is a sort of a bug that you could address or basically the answer is I cannot rely on `isDirty` when `shouldUnregister` is set to `false`?

**Comment by bluebill1049** [comment]
That's due to the `defaultListValue`, when `shouldUnregister: false` an input gets removed or unmount will fall back into the `defaultValues`, in your example 

Button 2 pressed: when [1, ] the second value will default back to the `defaultVlaue` which is `2`, so the current form value is [1, 2].
Button 1 pressed: when you get the form value to [2], the second value is fall back with `2`, so the current form value is [2, 2]

basically `defaultValues` for `shouldUnregister: false` is the fallback value when the input or value gets removed or unmounted.  

**Solution:** 

keep your data structure consistent, `[1, 2]` after press two `[1, null]`, or press one `[null, 2]`

eg: https://codesandbox.io/s/react-hook-form-controller-template-forked-hdzkw?file=/src/index.js

**Comment by bluebill1049** [comment]
<img width="938" alt="Screen Shot 2020-11-17 at 10 18 03 pm" src="https://user-images.githubusercontent.com/10513364/99383955-c96e2d80-2922-11eb-8dec-950d7ad400c2.png">

include some more information in the doc, let me know if it makes sense or to be improved.

**Comment by bluebill1049** [comment]
let me know if the above answer your question?

**Comment by bluebill1049** [comment]
Think i found a solution for this.

**Comment by bluebill1049** [comment]
next patch should solve this issue.

**Comment by Mhal007** [comment]
@bluebill1049 Sorry for the delay.

I have checked your proposal and gave it a shot.

At first, I've implemented replacing removed elements with `null`:
https://codesandbox.io/s/react-hook-form-controller-template-forked-7j8nr?file=/src/index.js

This has a couple of problems. For example, if you:
1. Remove 1
2. Remove 2
3. Add 1
4. Add 2

The state is `[null, null, 1, 2]`, and so the form is still dirty, even though it shouldn't.

So then I thought, ok, let's fix that. And so I've implemented it this way:
https://codesandbox.io/s/react-hook-form-controller-template-forked-oproj?file=/src/index.js

When you add a number I'm checking whether this value was included in the default array, and if so, placing it in its spot, if not, I'm adding it at the end. For removing, if an element is one of the defaultValues, then I replace it with `null`, if not, I just remove it from the current value array.

When you play around with it, it does work fine if the length of the array is constant, so you can add and remove 1 and 2 as much as you'd like, and it all works fine. But then, I realized that's not exactly my case. User can add more elements than just these of the default value, and so I've added the `ADD 3` button. And that, again, breaks the thing. If you Add 3, and then Remove 3, I'm not sure what needs to be done to get the `isDirty` property to go back to `false`. Within the link you can see three approaches: removing the '3' from the array, replacing it with `null`, replacing it with `undefined`, non of these work.

Basically, it's getting more and more complicated and convoluted for something that I would assume should work out of the box with just using `onChange(newValue)`.

Let me know what you think.

**Comment by Mhal007** [comment]
@bluebill1049 Just noticed you mentioned it'd be fixed with the new release :) So you can ignore the post above, or you can use these links for additional testing if you feel like it'd help.

Thanks!

**Comment by Mhal007** [comment]
@bluebill1049 I've seen that released two new patches. Was one of it suppose to fix the initial issue already? I've checked it with the newest version (6.11.5) and seems like it's not fixed yet.

**Comment by bluebill1049** [comment]
hey @Mhal007 i did test your example before the release: https://codesandbox.io/s/react-hook-form-controller-template-forked-qdf1e?file=/src/index.js let me know if your use case is changed, and please supply the steps to reproduce if not fixed.

**Comment by Mhal007** [comment]
Ah, yes, it does work :thinking: Must have done something wrong.

Works like a charm, thanks for your help! :)
